### PR TITLE
WIP: Flake pinning

### PR DESCRIPTION
**Pull request by hugosenari**
These are my tests trying to use nix flake for pinning.

**PR isn't ready yet, but some packages are working.**

It contains mostly tree parts:
1. refresher - nims code to index repositories from packages.json and create packages flake.nix.
2. packages flakes (ie: x/xiuaua/flake.nix, x/xiuaua/main/flake.nix, x/xiuaua/v0_1_0/flake.nix ...)
3. lib that ins't in this repository only for performance reasons, are in [nim-flakes-lib](https://github.com/riinr/nim-flakes-lib)

**Status**

- [x] Creates a nix package for every package in packages.json
- [ ] Creates a nix package for every package dependency that wasn't in package.json
- [x] Index meta information for every package. (meta.json)
- [x] Resolves dependencies using itself
- [x] Multiple dependencies resolution A ⇒(B, C, D)
- [ ] Sub dependencies resolution A⇾B⇾C
- [x] Dependency version resolutions as tags: `>=`, `<=`, `>`, `<`, `=`, `> && <`, `*`, `#main`
- [ ] Dependency type: `^=`, `~=`
- [x] system x86_64-linux
- [ ] other systems 

**Things I tried but didn't worked**
import with ../ in sub flake
it works for a package, but not for a package dependency: A depends B depends C (works for A ⇾ B but not for B ⇾ C)

If I try to create a lock nix generate a HASH for all files in repository, that's why lib is in other project. Because for every flake it redo it and it was taking too much time. In fact, the actual lock file was generated adding a sub git repository for every package.

I'm not sure why `nimpkgs/f/fugitive/master` works but `nimpkgs/f/fugitive` won't resolve dependencies, since lib just take master and reexpose as defaulPackage.

### CODE REVIEW

**Comment by hugosenari** [comment]
@ehmry sorry to bother you, but any tip with sub dependency resolution problem?

**Comment by ehmry** [comment]
> +7,715,193 −4

Absolutely not.

**Comment by dvn0** [comment]
@hugosenari - I attempted to review this by cloning your branch locally, and while it (incredibly) didn't cause my tools to crash, it was impossible to review as it's too large and your commit messages don't provide much context for a given commit. Can you try at least explaining a bit more what this PR is supposed to change, besides adding an insane amount of metadata.

**Comment by dvn0** [comment]
I don't expect @ehmry will change his mind, but I'm curious, nevertheless.

**Comment by ehmry** [comment]
I'm uncomfortable with how much metadata we are piling relative to how much this repo is used. I would prefer to keep all the packages in a top-level set in the flake, and perhaps the previous versions of a package as attrs on the current version, like `flake.packages.${system}.foo.passthru.older."0.0.1"`.

My frustration with Nimble packaging is that none of its problems are not unique. I would prefer that we dicuss any major undertaking with the rest of the Nix community, and do it in a way that is eventually consistent with the other languages.

**Comment by hugosenari** [comment]
Hi, thanks for your feedback.

I was expecting to use more from actual project but it ended mostly like a project fork. So I wasn't expecting it to be merged in its current state (or maybe never).

The best way to discuss this kind of change was an PR because it isn't a issue and with PR I could provide an POC. Sorry to bother you. 

The initial idea was to make it simpler but ended not that simple, it was to let nix flake take care of hashes by itself.

But a big flake with all packages wouldn't scale or initial lock could take ages. That's why I created 1 flake for every package. The metadata was just for 'caching'. I started with fewer files but relative (../) path didn't work.
The file creation was done by [updateFlake ](https://github.com/riinr/flake-nimble/blob/flake-pinning/nimpkgs/refresher/updatePkg.nims). It fetchs branches and tags saves metadata and flake.nix

Then I have to deal with different versions, so every package has an 'main' and subpackages for its versions (mostly git tags). Every dependency points to main packages and resolve it's required version. The actual resolution is done by the [lib](https://github.com/riinr/nim-flakes-lib/blob/master/lib.nix#L178),

In summary:
- it's nimbles2flakes instead of nimbles2flake
- most of changes are in /nimpkgs folder
- /nimpkgs/refresher/updateFlake has the code to file creation
- [a-z]/[pkgName]/flake.nix points to [a-z]/[pkgName]/[version]/flake.nix
- [a-z]/[pkgName]/[version]/flake.nix points to SRC and [a-z]/[otherPkgName]/flake.nix
- [lib](https://github.com/riinr/nim-flakes-lib/blob/master/lib.nix#L178) has de version resolution algorithm

Another option would be maintain a ~2500 micro repositories, but I not sure I where could host/manage this.

**Comment by hugosenari** [comment]
Instead of open another PR with a lot of nonsense, I will [repost](https://github.com/nix-community/dream2nix/discussions/418) it here:

Another [POC](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/example/flake.nix#L14) now works.
Documenting here, if anyone else would like to try or backport.

Example requesting 3 packages,
hcparse has [11 dependencies](http://www.plantuml.com/plantuml/uml/xPLH3i8W38RVFGKNy0YQ7S4hC22ZEWgX9OintbrmSs5UO9ZunDylVVXI6kY8PUJaneCG6eOeWnMNso0UkLOUbK6Eais9Op8PMHLa92EEvq_yWWv85LJE0iYQSQp1bJpUQz0RL7P34eBoskY-Xawgw1F3I3MTWEtM0C5bO-3IjkOTrROEkPB4fMrEpOY5407VOdFQyp8hWuRcLMlfA2V0eT0bjh7yJSshqc9QtuebDYBsN_iVNVldeJEeyx_r0W==)
libp2p has [16 dependencies](http://www.plantuml.com/plantuml/uml/xLR1RiCW3BttA_Y17HPfkstg9-mF9aAyXeK0mewgTjg_3ogM4f8sgRPJr1EFvsTRnfQJBH9trBLgmrWbTuxRchqrE_Rbxy9e0PeS9tWXru6db2njOTd3QqG3gWLoKg3d8Wgag9tH9f0H1R8ciXr95UWU1qEd9H4W5Ra12VQU2qUkI-9Ue9Svby0TOdY8Y7BjkT9fzagQiqZBLhY39JDWRyYzoWVTGub6D9wBwD-BIk8kKwMDIkZPeXBjYbgLLoM5WhwmqoMOWlgtjXvFfkrTFj7eTmGdkP97Jj9eBncJCwLSSCZ2ZWHfY5lIZRGV70d90Mz3wSajzynCLt9aodNEqj6l0C-4O_46hJGA9Vs7QRXyGTZYwRbvFBKdujas9hhfnNmVzlkm_s7OlpTRq9N_lVe1)
cligen has 0 dependencies
```nix
dependencies = nimrevs.lib.srcs {
  inherit pkgs;
  nimPkgs = [
    "cligen"
    "hcparse"
    "libp2p"  
  ];
};
```
It can be built as:
```bash
nix show-derivation 'github:riinr/flake-crown?dir=pkgsRev/example'
nix build 'github:riinr/flake-crown?dir=pkgsRev/example'
nix build 'github:riinr/flake-crown?dir=pkgsRev/example'#nixpkgs2211 # fail :(
```

Pkg Resolution process:
- indexing time
    1. [index metadata as cudf](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/packages.cudf)
    2. index metadata as json ([example](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/codeberg.org/annaaurora/snekim/meta.json))
    3. [index cudf <-> json](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/packages.nix), because cudf has name (no `_`) and [version (must be int)](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/codeberg.org/annaaurora/snekim/meta.json#L216) restrictions
- buildtime
    1. [call aspcud](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/resolve.nix#L17) with a "request" (ie: cudf index + 'install: cligen, hcparse, libp2p >= 1')
    2. parse the response (i replaced `:` with `=` called it as TOML)
    3. use the cudf <-> json to find ref, rev and meta info
    4. fetchGit {ref, rev}

Index process:
1. fetch [packages.json](https://github.com/nim-lang/packages/blob/master/packages.json) (is like nixpkgs of Nim)
2. git ls-remote for every pkg
3. [clone every ref](https://github.com/riinr/flake-crown/blob/flake-pinning/refresher/updateRevs.nim#L108) (with --filter=blob:none --sparse --depth 1)
4. use nimble dump (only available in latest version)
5. write the output of dump to a json file

Notes and questions:
- There is a bug for [aspcud with large version numbers](https://github.com/potassco/aspcud/issues/11) (converting semver to int isn't lossless)
- This POC [didn't work for 22.11](https://github.com/NixOS/nixpkgs/issues/206333), I don't know why.
- I'm betting on aspcud precision, because it uses pure logic to resolve dependencies, [Opam](https://opam.ocaml.org/doc/External_solvers.html) used to recommend it. Other options? ;-)
- This POC doesn't propagate dependencies. It only downloads and adds to compiler path. `["--path:depA --path:depB"]`
